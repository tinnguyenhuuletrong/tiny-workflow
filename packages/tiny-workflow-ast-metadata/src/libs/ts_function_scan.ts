import {
  MethodDeclaration,
  Project,
  PropertyAccessExpression,
  SyntaxKind,
} from "ts-morph";

export function scan_expression_call(
  methodDeclaration: MethodDeclaration,
  methodName: string
) {
  const callExpressions = methodDeclaration.getDescendantsOfKind(
    SyntaxKind.CallExpression
  );

  // Filter the call expressions to find those like `this.withAction`
  const withActionCalls = callExpressions.filter((callExpr) => {
    const expression = callExpr.getExpression();

    // Ensure that the expression is a PropertyAccessExpression
    if (expression.getKind() === SyntaxKind.PropertyAccessExpression) {
      const propertyAccess = expression as PropertyAccessExpression;

      // Check if the expression matches `this.withAction`
      return (
        // propertyAccess.getExpression().getText() === "this" &&
        propertyAccess.getName() === methodName
      );
    }

    return false;
  });

  return withActionCalls;
}
