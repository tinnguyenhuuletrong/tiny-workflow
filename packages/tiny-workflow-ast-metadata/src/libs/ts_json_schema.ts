import { type Type, type Node, Project } from "ts-morph";

export function convertTypeToSchema(type: Type, node: Node): any {
  const schema: any = {};

  if (type.isString()) {
    schema.type = "string";
  } else if (type.isNumber()) {
    schema.type = "number";
  } else if (type.isBoolean()) {
    schema.type = "boolean";
  } else if (type.isArray()) {
    schema.type = "array";
    schema.items = convertTypeToSchema(type.getArrayElementTypeOrThrow(), node);
  } else if (type.isObject()) {
    schema.type = "object";
    schema.properties = {};
    const properties = type.getProperties();
    for (const prop of properties) {
      const propType = prop.getTypeAtLocation(node);
      schema.properties[prop.getName()] = convertTypeToSchema(propType, node);
    }
  } else if (type.isUnion()) {
    const unionVal = type.getUnionTypes();
    schema.type = "string";
    schema.enum = unionVal.map((itm) => itm.getText().slice(1, -1));
  } else {
    schema.type = "unknown";
  }

  return schema;
}
