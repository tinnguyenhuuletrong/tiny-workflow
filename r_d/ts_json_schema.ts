import { type Type, type Node, Project } from "ts-morph";

export function convertTypeToSchema(type: Type, node: Node): any {
  const schema: any = {};

  if (type.isString()) {
    schema.type = "string";
  } else if (type.isNumber()) {
    schema.type = "number";
  } else if (type.isBoolean()) {
    schema.type = "boolean";
  } else if (type.isArray()) {
    schema.type = "array";
    schema.items = convertTypeToSchema(type.getArrayElementTypeOrThrow(), node);
  } else if (type.isObject()) {
    schema.type = "object";
    schema.properties = {};
    const properties = type.getProperties();
    for (const prop of properties) {
      const propType = prop.getTypeAtLocation(node);
      schema.properties[prop.getName()] = convertTypeToSchema(propType, node);
    }
  } else {
    schema.type = "unknown";
  }

  return schema;
}

async function main() {
  // Initialize the project
  const project = new Project({
    tsConfigFilePath: "tsconfig.json",
  });

  // Get the source file where the type is defined
  const sourceFile = project.getSourceFileOrThrow("examples/userOnboarding.ts");

  // Assuming you have a type alias like `type MyType = ...;`
  const typeAlias = sourceFile.getTypeAliasOrThrow("TStateShape");

  // Get the type from the alias
  const typeChecker = project.getTypeChecker();
  const type = typeChecker.getTypeAtLocation(typeAlias.getNameNode());

  const jsonSchema = convertTypeToSchema(type, typeAlias.getNameNode());
  console.log(JSON.stringify(jsonSchema, null, 2));
}

// main();
