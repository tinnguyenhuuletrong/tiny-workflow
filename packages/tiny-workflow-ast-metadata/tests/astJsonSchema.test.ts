import { test, expect, mock } from "bun:test";

import { convertTypeToSchema } from "../src/libs/ts_json_schema";
import { Project } from "ts-morph";

test("ast-jsonSchema-test", async () => {
  const T1 = `
  type SubT1 = {
    s1: {
      s11: number
    },
    s2: string
  }

  type T1 = {
    a: string,
    b: number,
    c: string[],
    d: number[],
    e: {
      e1: string,
      e2: 'e2_a' | 'e2_b'
    }
    f: SubT1
    g: SubT1[]
  }`;

  const project = new Project({});
  const file = project.createSourceFile("code.ts", T1);

  const typeAlias = file.getTypeAliasOrThrow("T1");
  const typeChecker = project.getTypeChecker();
  const type = typeChecker.getTypeAtLocation(typeAlias.getNameNode());

  const jsonSchema = convertTypeToSchema(type, typeAlias.getNameNode());
  expect(jsonSchema).toMatchSnapshot();
});
