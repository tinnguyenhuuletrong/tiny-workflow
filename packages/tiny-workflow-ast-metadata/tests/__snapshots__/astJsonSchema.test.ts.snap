// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ast-jsonSchema-test 1`] = `
{
  "properties": {
    "a": {
      "type": "string",
    },
    "b": {
      "type": "number",
    },
    "c": {
      "items": {
        "type": "string",
      },
      "type": "array",
    },
    "d": {
      "items": {
        "type": "number",
      },
      "type": "array",
    },
    "e": {
      "properties": {
        "e1": {
          "type": "string",
        },
        "e2": {
          "enum": [
            "e2_a",
            "e2_b",
          ],
          "type": "string",
        },
      },
      "type": "object",
    },
    "f": {
      "properties": {
        "s1": {
          "properties": {
            "s11": {
              "type": "number",
            },
          },
          "type": "object",
        },
        "s2": {
          "type": "string",
        },
      },
      "type": "object",
    },
    "g": {
      "items": {
        "properties": {
          "s1": {
            "properties": {
              "s11": {
                "type": "number",
              },
            },
            "type": "object",
          },
          "s2": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;
